src/rest/resource.cpp

```diff
@@ -800,6 +809,252 @@ void Resource::DatasetPending(const Request &aRequest, Response &aResponse) cons
+{
+    std::string errorCode;
+
+    switch (aRequest.GetMethod())
+    {
+    case HttpMethod::kGet:
+        GetCommissionerState(aResponse);
+        break;
+    case HttpMethod::kPut:
+        SetCommissionerState(aRequest, aResponse);
+        break;
+    case HttpMethod::kOptions:
+        errorCode = GetHttpStatus(HttpStatusCode::kStatusOk);
+        aResponse.SetResponsCode(errorCode);
+        aResponse.SetComplete();
+        break;
+    default:
+        ErrorHandler(aResponse, HttpStatusCode::kStatusMethodNotAllowed);
+        break;
+    }
+}
+
+void Resource::GetJoiners(Response &aResponse) const
+{
+    uint16_t                  iter = 0;
+    otJoinerInfo              joinerInfo;
+    std::vector<otJoinerInfo> joinerTable;
+    std::string               joinerJson;
+    std::string               errorCode;
+
+    while (otCommissionerGetNextJoinerInfo(mInstance, &iter, &joinerInfo) == OT_ERROR_NONE)
+    {
+        joinerTable.push_back(joinerInfo);
+    }
+
+    joinerJson = Json::JoinerTable2JsonString(joinerTable);
+    aResponse.SetBody(joinerJson);
+    errorCode = GetHttpStatus(HttpStatusCode::kStatusOk);
+    aResponse.SetResponsCode(errorCode);
+}
+
+void Resource::AddJoiner(const Request &aRequest, Response &aResponse) const
+{
+    otbrError           error   = OTBR_ERROR_NONE;
+    otError             otError = OT_ERROR_NONE;
+    std::string         errorCode;
+    otJoinerInfo        joiner;
+    const otExtAddress *addrPtr                         = nullptr;
+    const uint8_t       emptyArray[OT_EXT_ADDRESS_SIZE] = {0};
+
+    VerifyOrExit(otCommissionerGetState(mInstance) == OT_COMMISSIONER_STATE_ACTIVE, error = OTBR_ERROR_INVALID_STATE);
+
+    VerifyOrExit(Json::JsonJoinerInfoString2JoinerInfo(aRequest.GetBody(), joiner), error = OTBR_ERROR_INVALID_ARGS);
+
+    addrPtr = &joiner.mSharedId.mEui64;
+    if (memcmp(&joiner.mSharedId.mEui64, emptyArray, OT_EXT_ADDRESS_SIZE) == 0)
+    {
+        addrPtr = nullptr;
+    }
+
+    if (joiner.mType == OT_JOINER_INFO_TYPE_DISCERNER)
+    {
+        otError = otCommissionerAddJoinerWithDiscerner(mInstance, &joiner.mSharedId.mDiscerner, joiner.mPskd.m8,
+                                                       joiner.mExpirationTime);
+    }
+    else
+    {
+        otError = otCommissionerAddJoiner(mInstance, addrPtr, joiner.mPskd.m8, joiner.mExpirationTime);
+    }
+
+    VerifyOrExit(otError != OT_ERROR_NO_BUFS, error = OTBR_ERROR_OPENTHREAD);
+    VerifyOrExit(otError != OT_ERROR_INVALID_ARGS, error = OTBR_ERROR_INVALID_ARGS);
+
+    errorCode = GetHttpStatus(HttpStatusCode::kStatusOk);
+    aResponse.SetResponsCode(errorCode);
+
+exit:
+    if (error == OTBR_ERROR_INVALID_STATE)
+    {
+        ErrorHandler(aResponse, HttpStatusCode::kStatusConflict);
+    }
+    else if (error == OTBR_ERROR_INVALID_ARGS)
+    {
+        ErrorHandler(aResponse, HttpStatusCode::kStatusBadRequest);
+    }
+    else if (error == OTBR_ERROR_OPENTHREAD)
```

Comment:
Instead of using OTBR_ERROR_OPENTHREAD for InsufficientStorage, it's probably better to just use OT error directly, instead of casting OT error to OTBR error.